### ISSUE TITLE
Type declarations do not understand typed-function's automatic conversions

### ISSUE THREAD

**Issue by gwhitney** [josdejong/mathjs#2582]

**Comment by josdejong** [comment]
Thanks for bringing this up. Looking at the definition of `exp` for example, it looks like:

```ts
    /**
     * Calculate the exponent of a value. For matrices, the function is
     * evaluated element wise.
     * @param x A number or matrix to exponentiate
     * @returns Exponent of x
     */
    exp(x: number): number
    exp(x: BigNumber): BigNumber
    exp(x: Complex): Complex
    exp(x: MathArray): MathArray
    exp(x: Matrix): Matrix
```

This looks well defined, only, `Fraction` is missing. I see `Fraction` is missing for many functions. I think there is not directly a need to solve this in a generic way or so, but we can just go the dumb way and add type definitions for `Fraction` where missing. Anyone able to help out here?

**Comment by josdejong** [comment]
I understand your point. Right now, mathjs is plain JavaScript, and the TypeScript definitions do not fully cover mathjs. They come "close". It is not trivial to rewrite mathjs in TypeScript (though I would like to take that step). So until then, we simply have to deal with the current situation. It will not be possible to make the types 100% right. I guess there is a sweat spot somewhere that we have it 80% or 90% correct with "relatively little effort". Trying to get towards 100% will cost so much effort that it is not worth it. 

Improving the types for `Fraction` the dumb way by writing the types by hand will probably take 15-30 minutes, so I guess that is worth it. The more subtle and complex conversions and differing behavior depending on config options like `predictable` is probably too much. What do you think?

**Comment by mattvague** [comment]
>I understand your point. Right now, mathjs is plain JavaScript, and the TypeScript definitions do not fully cover mathjs. They come "close". It is not trivial to rewrite mathjs in TypeScript (though I would like to take that step). So until then, we simply have to deal with the current situation. It will not be possible to make the types 100% right. I guess there is a sweat spot somewhere that we have it 80% or 90% correct with "relatively little effort". Trying to get towards 100% will cost so much effort that it is not worth it.

@josdejong as I understood on our call, the biggest (or first?) hurdle would be trying to get https://github.com/josdejong/typed-function to work with Typescript, is that right? If so, I have a few different ideas and would be happy to start spiking on that a bit.

Some references / possible starting points:
- https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html
- https://github.com/gcanti/io-ts

**Comment by mattvague** [comment]
Is there already a task / discussion for the TS conversion?

**Comment by josdejong** [comment]
Yes correct, first step would be to add TypeScript support to `typed-function`. There is no single TypeScript topic discussing though we've had discussions on it here and there. Let's use #2076 to further discuss it ok?

**Comment by gwhitney** [comment]
Thanks for moving the discussion on converting mathjs to TypeScript elsewhere, as this issue seems to be primarily about being able to correctly _use_ mathjs from TypeScript. And in response to that, I just wanted to say that as far as I can see, even ignoring features like the 'predictable' parameter (which we could just ban the use of from TypeScript), the current set of signatures isn't anything like 90%. It's like 90% if you ignore all type conversions -- the TypeScript declarations seem to generally be derived from the calls to typed, which _of course_ do not mention all of the convertible types, that's the point of automatic conversion. But there are a dozen or more automatic conversions in mathjs, and the vast majority of typed functions in mathjs potentially trigger one or more of them, so taking those conversions into account, the current typescript signatures are more like 10% correct (although in practice a very useful 10%!).

I'll be frank, I think the best way forward here would be to extend typed-function to allow return-type annotations as well (which has other uses and would also be an incremental step toward converting to TypeScript anyway) and then add a method to a typed function which emits a valid TypeScript declaration for it. Then we could ditch (most of) the current index.d.ts and just generate it at build time.

Short of that, the only method I see for getting index.d.ts close to right is to define types like 'ConvertibleToNumber' which is a union of number and all types for which there is an automatic conversion to number and use them in place of 'number' in most instances in index.d.ts (and the same for all other types which are the destinations of conversions) and try it and see if it all works out.

**Comment by josdejong** [comment]
> the current typescript signatures are more like 10% correct

ha ha now you almost make me cry 😉 

Agree, types in `typed-function` should be addressed first. And then we need a solution to auto-magically generate all these types (including based on conversions etc) for us. That will still not address things like differing behavior based on config like `predictable`, but at least get us much further than 10%.
